### ISSUE TITLE
Support for multiprocessing within PyTorch data loaders

### ISSUE THREAD

**Issue by btgraham** [getkeops/keops#45]
Hello. Can pykeops be used in subprocesses, i.e. in PyTorch DataLoader methods, please?
```python
import torch
import pykeops.torch

def nearest(a,b):
    print(a.shape,b.shape)
    a=a[:,None,:]
    b=b[None,:,:]
    a= pykeops.torch.LazyTensor(a)
    b= pykeops.torch.LazyTensor(b)
    return ((a-b)**2).sum(2).argmin(1).flatten()

class dataset(torch.utils.data.Dataset):
    def __len__(self):
        return 10
    def __getitem__(self,k):
        return nearest(torch.randn(10,3),torch.randn(10,3))

for x in torch.utils.data.DataLoader(dataset(), batch_size=None, num_workers=0):
    print(x)
#This works with num_workers==0
    
for x in torch.utils.data.DataLoader(dataset(), batch_size=None, num_workers=1):
    print(x)
#This fails for num_workers==1
# RuntimeError: DataLoader worker (pid(s) 6743) exited unexpectedly
```

**Comment by bcharlier** [comment]
Hi @btgraham ,

I do not know the DataLoader class... but it seems to rely on Python multiprocessing. This is a good news, since I am pretty sure that you can use pyKeOps in subprocesses as discussed [here](https://github.com/getkeops/keops/issues/27#issuecomment-540045715). Unfortunately, as far as I know,  you will have to tune some parameters and/or write a wrapper around pykeops code...


b.

**Comment by joanglaunes** [comment]
Hello  @btgraham ,
I guess this comes too late in your case, but I have found a way to make your example work. 
A first remark is that this is linked to the use of the Gpu. Your script will work if Gpu is disabled by adding the option `backend='CPU'` in the call to `argmin`.
Now for the default Gpu mode, as can be seen in @bcharlier link, one needs to set the "spawn" method instead of the "fork" start method for multiprocessing, when using the Gpu. This is said also in PyTorch documentation here :
https://pytorch.org/docs/stable/notes/multiprocessing.html#multiprocessing-cuda-note
Now here is the working code :
```python
import torch
import pykeops.torch

from multiprocessing import set_start_method
try:
    set_start_method('spawn')
except RuntimeError:
    pass

def nearest(a,b):
    print(a.shape,b.shape)
    a=a[:,None,:]
    b=b[None,:,:]
    a= pykeops.torch.LazyTensor(a)
    b= pykeops.torch.LazyTensor(b)
    return ((a-b)**2).sum(2).argmin(1).flatten()

class dataset(torch.utils.data.Dataset):
    def __len__(self):
        return 10
    def __getitem__(self,k):
        return nearest(torch.randn(10,3),torch.randn(10,3))

if __name__ == '__main__':
	for x in torch.utils.data.DataLoader(dataset(), batch_size=None, num_workers=1):
		print(x)
```
Note that I had also to add a  `if __name__ == '__main__'` statement, this is apparently required when using the multiprocessing package. Also this example works only as a script called from a terminal and not inside a Jupyter notebook. This problem with multiprocessing package and Jupyter notebooks is also known apparently as I found references to it on the internet... 
A last remark as that if you want to use directly PyTorch Gpu tensors in the dataloader (instead of using KeOps on Cpu tensors as input and output, with Gpu support in the middle), then it will fail, and this is not linked to KeOps! It will fail also if you use pure PyTorch code. There is a warning about this in PyTorch documentation :
https://pytorch.org/docs/stable/data.html#multi-process-data-loading

If you have more experience on the use of Python multiprocessing and the Dataloader class, let us know. I'm still thinking how to include a patch, warning or explicit error in KeOps code to close this issue.

Joan
